### ISSUE TITLE
Multiple matching crates error, if dependency on compiletest is optional

### ISSUE THREAD

**Issue by hannobraun** [laumann/compiletest-rs#114]
I'm trying to use compiletest for a `#[no_std]` crate. Because the crate contains examples that need to be built for an embedded platform where `std` is not available (thumbv6m-none-eabi), I needed to make the dependency on compiletest optional and add a Cargo feature for it.

Further details:
- My [Cargo.toml](https://github.com/hannobraun/rust-lpc82x-hal/blob/4ae58692fe45a0ef52b960e965b9dd1712caa572/Cargo.toml) (important parts: [1](https://github.com/hannobraun/rust-lpc82x-hal/blob/4ae58692fe45a0ef52b960e965b9dd1712caa572/Cargo.toml#L28), [2](https://github.com/hannobraun/rust-lpc82x-hal/blob/4ae58692fe45a0ef52b960e965b9dd1712caa572/Cargo.toml#L37))
- My [test file](https://github.com/hannobraun/rust-lpc82x-hal/blob/4ae58692fe45a0ef52b960e965b9dd1712caa572/tests/compiletest.rs)

This works works fine:
```
$ cargo clean
$ cargo test --features compiletest
```

This does not work:
```
$ cargo test
$ cargo test --features compiletest
```

It results in the following error (excerpt from [full build log](https://travis-ci.org/braun-robotics/rust-lpc82x-hal/builds/382540324)):
```
status: exit code: 101
command: "rustc" "tests/compile-fail/swm/assign-function-to-pin-with-unknown-state.rs" "-L" "/tmp" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/swm/assign-function-to-pin-with-unknown-state.stage-id" "-L" "/home/travis/build/braun-robotics/rust-lpc82x-hal/target/debug/build/cortex-m-rt-e3337e5285e3af86/out" "-L" "/home/travis/build/braun-robotics/rust-lpc82x-hal/target/debug" "-L" "/home/travis/build/braun-robotics/rust-lpc82x-hal/target/debug/deps" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib" "-L" "/tmp/swm/assign-function-to-pin-with-unknown-state.stage-id.aux" "-A" "unused"
unexpected errors (from JSON output): [
    Error {
        line_num: 1,
        kind: Some(
            Error
        ),
        msg: "1:1: 1:25: multiple matching crates for `lpc82x_hal` [E0464]"
    },
    Error {
        line_num: 1,
        kind: Some(
            Error
        ),
        msg: "1:1: 1:25: found crate `lpc82x_hal` compiled by an incompatible version of rustc [E0514]"
    }
]
```

I believe the important part is `multiple matching crates`. I think I understand what happens here: We have two versions of the crate, one with the feature, one without, and rustc doesn't know which one to use. Is this something that could be fixed in `clean_rmeta`?

**Comment by laumann** [comment]
Yes, this *should* be something that's fixable by `clean_rmeta()`, although that function was created to deal with `cargo check` followed by `cargo test`. I wonder if this is the same issue as #101

**Comment by laumann** [comment]
Hmm, looking at [`clean_rmeta()`](https://github.com/laumann/compiletest-rs/blob/a0f97ccc46512eb959fefffa2ef64d29e59bc1ef/src/common.rs#L258-L280), could it be that the filter:

```
            .filter(|s| s.ends_with("/deps"))
```
is the problem?

**Comment by laumann** [comment]
You might also want to try not using `link_deps()` and just set `config.target_rustcflags` manually.

**Comment by hannobraun** [comment]
Thanks for your replies! Just FYI, I won't be able to follow up here today, and tomorrow I'm off to RustFest. I'll check back in next week, or the week after.

**Comment by laumann** [comment]
No problem :smile: This is a problem that's been a thorn in my side for some time, so it'd be nice to have it solved!

Have fun at RustFest! Say hi to the guys from Seasoned Software if you catch them!

**Comment by hannobraun** [comment]
> Have fun at RustFest! Say hi to the guys from Seasoned Software if you catch them!

Thank you! I don't know who they are, but I'll keep my eyes open :)

**Comment by laumann** [comment]
Thanks for looking into this.

Yes, you are correct, `link_deps()` just adds folders for rustc to find libraries. Usually you'd just set `target_rustcflags` to `-L target/debug` or something (the folder where compiled libs are placed).

Sounds like the rlibs might be the problem, as you say. When `cargo test` fails, do you have multiple versions of the same `rlibs`? What happens if you then try to specifically remove all rlibs before another `cargo test`?

**Comment by hannobraun** [comment]
> When cargo test fails, do you have multiple versions of the same `rlibs`?

Yes. After a `cargo clean && cargo test && cargo test --features compiletest`, I have these two rlibs in `target/debug/deps`:
```
liblpc82x_hal-9e595ce07b7a25b7.rlib
liblpc82x_hal-d597ff95b0e3d874.rlib
```

And another one in `target/debug`:
```
liblpc82x_hal.rlib
```

The first `cargo test` creates the rlib in `target/debug` and one in `target/debug/deps`. The `cargo test --features compiletest` adds the second one in `target/debug/deps`.

Interestingly, if I run just `cargo clean && cargo test --features compiletest`, it also creates the rlib in `target/debug`, in addition to the one in `target/debug/deps`.

> What happens if you then try to specifically remove all rlibs before another `cargo test`?

If I remove the rlibs (and just the rlibs), both variants of `cargo test` succeed after that.

**Comment by laumann** [comment]
So we can pretty much say that the "duplicate" `rlib`s are the problem. Question is then, how do we fix it?

Can we identify which is which? If so, we might be able to just add those as link flags, instead of the entire folder. Otherwise I'm not what to do.

Surely, Cargo has a way of knowing which is which?

**Comment by hannobraun** [comment]
> Can we identify which is which? If so, we might be able to just add those as link flags, instead of the entire folder. Otherwise I'm not what to do.
>
> Surely, Cargo has a way of knowing which is which?

I've taken a look at the Cargo source trying to find that out, but didn't get very far before I ran out of time. Certainly it must know, but the question is how to get that knowledge out of Cargo and into compiletest.

I plan to take a closer look at the Cargo code next week (if I find the time).

**Comment by laumann** [comment]
OK, thanks again for looking into it :smiley:

**Comment by hannobraun** [comment]
I've taken another look at the Cargo code, but didn't come up with anything useful before running out of time. I've now added a workaround on my side (disabled caching on Travis), and have decided to leave this be for now.

Sorry, I would have liked to fix this properly, but my time is limited and I have to pick my battles.

**Comment by MoSal** [comment]
meh, this is kind of annoying. One has to always `rm -rf target` before running `cargo test` to make
sure this bug will not be hit.

**Comment by aldanor** [comment]
Just ran into this as well (#147). The `.clean_rmeta()` indeed removes `target/debug/deps/*.rmeta` files, however that doesn't seem sufficient, and you still run into both E0463 (crate not found) and E0464 (multiple matching crates) errors, sometimes both at the same time, somehow.

Has anyone figured out what exactly needs to be cleaned so it works in a stable fashion?

**Comment by aldanor** [comment]
@laumann Ok, here's what seems to work for me:

1. For each `foo.rmeta` file, also kill `foo.*` in the same folder (e.g., `foo.d`).
2. Don't filter by `*/deps`; do it in all link paths.

Using the cleanup code below (to replace `.clean_rmeta()`), I can now run `cargo check` or `cargo clippy` followed by `cargo test` and it wouldn't choke on E0463 / E0464. 

Disclaimer: there may be some gotchas I'm not aware about but it seems to work so far for me.

```rust
use std::fs::{read_dir, remove_file};

fn clean_rlibs(config: &compiletest_rs::Config) {
    if config.target_rustcflags.is_some() {
        for directory in config.target_rustcflags
            .as_ref()
            .unwrap()
            .split_whitespace()
            {
                if let Ok(mut entries) = read_dir(directory) {
                    while let Some(Ok(entry)) = entries.next() {
                        let f = entry.file_name().clone().into_string().unwrap();
                        if f.ends_with(".rmeta") {
                            let prefix = &f[..f.len() - 5];
                            let _ = remove_file(entry.path());
                            if let Ok(mut entries) = read_dir(directory) {
                                while let Some(Ok(entry)) = entries.next() {
                                    let f = entry.file_name().clone().into_string().unwrap();
                                    if f.starts_with(prefix) && !f.ends_with(".rmeta") {
                                        let _ = remove_file(entry.path());
                                    }
                                }
                            }
                        }
                    }
                }
            }
    }
}
```

**Comment by aldanor** [comment]
@laumann Np. Yea, this is a replacement to the current `.clean_rmeta()`, basically it just removes more files.

Disclaimer: I don’t have deep understanding of why this works, I just took a snapshot of `target` folder before and after running a cargo check/clippy and removed all the new files that appeared. That being said, this has been working perfectly fine for me the last few days with no issues.

I could open a PR if that helps :)

**Comment by laumann** [comment]
@aldanor A PR would be nice - then we could ask others for feedback. I tried it out on the test-project and it seems to work fine.

**Comment by aldanor** [comment]
@laumann #148

**Comment by orium** [comment]
I've avoided this issue by mimicking the way cargo calls `rustc`.  I set the rustc flags like this:

```rust
config.target_rustcflags = Some("--edition=2021 -L dependency=target/debug/deps/ --extern mycrate=<path to mycrate rlib> --extern adependency=<path to adependency rlib>");
```

where `mycrate` is the name of your crate and `adependency` is a crate you depend on and want to access on your tests.

You can see an example of this working [here](https://github.com/orium/archery/commit/da28e3500ce168db11b3539cb2b2077641fd7ad0#diff-2ad10836ccce5ac2056d5679cc92449d9ff9094d4ff5c5803f65b5dd1d52ef19).
